sympy/solvers/solvers.py

```diff
@@ -2653,22 +2655,58 @@ def _tsolve(eq, sym, **flags):
             elif sym not in lhs.exp.free_symbols:
                 return _solve(lhs.base - rhs**(1/lhs.exp), sym, **flags)
-            elif not rhs and sym in lhs.exp.free_symbols:
-                # f(x)**g(x) only has solutions where f(x) == 0 and g(x) != 0 at
-                # the same place
-                sol_base = _solve(lhs.base, sym, **flags)
-                if not sol_base:
-                    return sol_base  # no solutions to remove so return now
-                return list(ordered(set(sol_base) - set(
-                    _solve(lhs.exp, sym, **flags))))
-            elif (rhs is not S.Zero and
-                        lhs.base.is_positive and
-                        lhs.exp.is_real):
-                return _solve(lhs.exp*log(lhs.base) - log(rhs), sym, **flags)
-            elif lhs.base == 0 and rhs == 1:
-                return _solve(lhs.exp, sym, **flags)
             else:
-                raise NotImplementedError
+                # sym in lhs.exp.free_symbols
+                if not rhs:
+                    # f(x)**g(x) only has solutions where f(x) == 0 and g(x) != 0 at
+                    # the same place
+                    sol_base = _solve(lhs.base, sym, **flags)
+                    if not sol_base:
+                        return sol_base  # no solutions to remove so return now
+                    return list(ordered(set(sol_base) - set(
+                    _solve(lhs.exp, sym, **flags))))
+                elif rhs.is_Rational or lhs.exp.is_real:
+                    sol=[]
+                    if lhs.base.is_real and lhs.base<0:
+                        sol.extend(_solve(-abs(lhs.base)**lhs.exp - rhs, sym, **flags) + _solve(abs(lhs.base)**lhs.exp - rhs, sym, **flags))
+                    if lhs.base.is_positive:
+                        if lhs.exp.is_real:
+                            sol.extend(_solve(lhs.exp*log(lhs.base) - log(rhs), sym, **flags))
+                        else:
+                            rewrite = lhs.rewrite(exp)
+                            if rewrite != lhs:
+                                sol.extend(_solve(rewrite - rhs, sym, **flags))
+                    if rhs.is_Rational:
+                        b_pos=[]
+                        b_negi=[]
+                        e0=[]
+                        for i in divisors(abs(rhs.p)):
+                            b=0
+                            t=True
+                            if i!=1:
+                                b, t = integer_log(rhs.p,i)
+                            if t:
+                                for s in divisors(abs(rhs.q)):
+                                    if s**b== rhs.q:
+                                        b_pos.extend(_solve(lhs.base-Rational(i,s), sym, **flags))
+                                        b_negi.extend(_solve(lhs.base+Rational(i,s), sym, **flags))
+                                        e0.extend(_solve(lhs.exp-b, sym, **flags))
```

Comment:
This should be factored out because `b` (or `e` if you change the name) does not depend on `s`...and you don't want to keep solving the same equation over and over.

Reply:
Okk I will review and edit as you suggested.